experimental/ir/option.go

```diff
@@ -0,0 +1,308 @@
+	"math"
+
+	"github.com/bufbuild/protocompile/experimental/ast"
+	"github.com/bufbuild/protocompile/experimental/ast/predeclared"
+	"github.com/bufbuild/protocompile/experimental/internal"
+	"github.com/bufbuild/protocompile/experimental/seq"
+	"github.com/bufbuild/protocompile/internal/arena"
+	"github.com/bufbuild/protocompile/internal/ext/slicesx"
+	"github.com/bufbuild/protocompile/internal/intern"
+)
+
+// Option is an option in a Protobuf file.
+type Option struct {
+	withContext
+	raw *rawOption
+}
+
+type rawOption struct {
+	// This is an alternating sequence of path parts, where the first is
+	// a non-extension, the second is an extension, and so on.
+	//
+	// For example, foo.(bar.baz).bang.bonk is represented as
+	// ["foo", "bar.baz", "bang", "", "bonk"]. The empty string is used as a
+	// separator for adjacent components of the same type.
+	name []intern.ID //nolint:unused // Will become used in a followup.
+
+	field ref[rawField] // Set if this option matches a field.
+
+	// TODO: it may be worth inlining small integer values in the common case?
+	value arena.Pointer[rawValue]
+}
+
+// Field returns the message field that this option corresponds to.
+//
+// This is primarily used for type-checking; the path of fields from the
+// *Options descriptor at the root is not stored.
+func (o Option) Field() Field {
+	return wrapField(o.Context(), o.raw.field)
+}
+
+// Value returns the value this option is set to.
+func (o Option) Value() Value {
+	return wrapValue(o.Context(), o.raw.value)
+}
+
+func wrapOption(c *Context, p arena.Pointer[rawOption]) Option {
+	if c == nil || p.Nil() {
+		return Option{}
+	}
+	return Option{
+		withContext: internal.NewWith(c),
+		raw:         c.arenas.options.Deref(p),
+	}
+}
+
+// Value is an evaluated expression assigned to an [Option].
+type Value struct {
+	withContext
+	raw *rawValue
+}
+
+type rawValue struct {
+	ast  ast.ExprAny
+	bits rawValueBits
+	ty   predeclared.Name
+
+	isArray                  bool // If set, value is an arena.Pointer[[]rawValueBits].
+	isUninterpretedPath      bool // If set, value is an intern.ID.
+	isUninterpretedAggregate bool // If set, value is an intern.ID.
+}
+
+// rawValueBits is used to represent the actual value for all types, according to
+// the following encoding:
+//  1. All numeric types, including bool and enums. This holds the bits.
+//  2. String and bytes. This holds an intern.ID.
+//  3. Messages. This holds an arena.Pointer[rawMessage].
+//  4. Repeated fields. This holds an arena.Pointer[[]rawValueBits], where each
+//     value is interpreted as a non-array with this value's type.
+//     This exploits the fact that arrays cannot contain other arrays.
+type rawValueBits uint64
+
+// AST returns the expression that evaluated to this value.
+func (v Value) AST() ast.ExprAny {
+	return v.raw.ast
+}
+
+// Type is the element type of this value. For arrays, this is the type of the
+// element; Protobuf does not have a concept of an "array type".
```

Comment:
It may behoove us to provide an "array" type, even though Protobuf descriptors do not. That could enable simpler code that inspects and works with types, to treat them more uniformly. Though I think doing so requires adding another byte to `Type`, to put an `isArray` flag there 🤷 (which may not be desirable since, for alignment, it means likely adding 4 or 8 bytes to the `Type` struct in practice...).

Reply:
I am slightly skeptical of that. My gut instinct would be to not add that for now, and see how painful it makes things.